crates/ty_python_semantic/src/types/infer/builder/type_expression.rs

```diff
@@ -155,7 +155,12 @@ impl<'db> TypeInferenceBuilder<'db, '_> {
                     }
                     // anything else is an invalid annotation:
                     op => {
-                        self.infer_binary_expression(binary, TypeContext::default());
+                        // Avoid inferring the types of invalid binary expressions that have been
+                        // parsed from a string annotation, as they are not present in the semantic
+                        // index.
+                        if !self.deferred_state.in_string_annotation() {
+                            self.infer_binary_expression(binary, TypeContext::default());
+                        }
```

Comment:
We used to have the requirement that inference had to run for every expression in the AST, or otherwise we might panic when requesting a type for that expression. We have changed that and now fall back to `Unknown` in case no type has been inferred.

Since we're not going to infer useful types here anyway, I think we can also just remove this entirely.

Reply:
That leads to a worse experience for server users when they hover over sub-expressions in the invalid type expression, no? I think we should still aim to store a type for each sub-expression wherever possible

Reply:
I don't think that's really a realistic example of an invalid type expression. If somebody is using the appeal library, for example, they might be making use of that library's [special DSL](https://github.com/larryhastings/appeal?tab=readme-ov-file#annotations-and-introspection) for annotations (which is incompatible with type checkers, and so is going to result in lots of `invalid-type-form` violations). Who knows if this user even has `invalid-type-form` enabled in their ty config; maybe they're only interested in our server functionality. I don't see why the invalid type expression should mean that we don't offer any on-hover information or go-to-definition support for any symbols in this user's annotations.